// boneyard-js/dist/angular.js
import { Component, Input, ElementRef, ViewChild, ChangeDetectorRef, ChangeDetectionStrategy, } from '@angular/core';
import { CommonModule } from '@angular/common';
import { normalizeBone } from './types.js';
import { adjustColor, ensureBuildSnapshotHook, getRegisteredBones, isBuildMode, registerBones, resolveResponsive, SHIMMER, PULSE, DEFAULTS, } from './shared.js';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export { registerBones };
let _globalConfig = {};
export function configureBoneyard(config) {
    _globalConfig = { ..._globalConfig, ...config };
}
ensureBuildSnapshotHook();
export class SkeletonComponent {
    cdr;
    loading = false;
    name;
    initialBones;
    color;
    darkColor;
    animate;
    stagger = false;
    transition = false;
    boneClass;
    cssClass;
    snapshotConfig;
    containerRef;
    uid = Math.random().toString(36).slice(2, 8);
    buildMode = isBuildMode();
    containerWidth = 0;
    containerHeight = 0;
    isDark = false;
    activeBones = null;
    transitioning = false;
    get resolvedBoneClass() {
        return this.boneClass ?? _globalConfig.boneClass;
    }
    get staggerMs() {
        const v = this.stagger ?? _globalConfig.stagger;
        return v === true ? 80 : v === false || !v ? 0 : v;
    }
    get transitionMs() {
        const v = this.transition ?? _globalConfig.transition;
        return v === true ? 300 : v === false || !v ? 0 : v;
    }
    transitionTimer = null;
    resizeObserver = null;
    mutationObserver = null;
    mq = null;
    mqHandler = null;
    constructor(cdr) {
        this.cdr = cdr;
    }
    get resolvedColor() {
        return this.isDark
            ? (this.darkColor ?? _globalConfig.darkColor ?? DEFAULTS.web.dark)
            : (this.color ?? _globalConfig.color ?? DEFAULTS.web.light);
    }
    get animationStyle() {
        const raw = this.animate ?? _globalConfig.animate ?? 'pulse';
        return raw === true ? 'pulse' : raw === false ? 'solid' : raw;
    }
    get serializedSnapshotConfig() {
        return this.snapshotConfig ? JSON.stringify(this.snapshotConfig) : undefined;
    }
    get showSkeleton() {
        return (this.loading || this.transitioning) && !!this.activeBones;
    }
    get showFallback() {
        return this.loading && !this.activeBones && !this.transitioning;
    }
    get scaleY() {
        const effectiveHeight = this.containerHeight > 0 ? this.containerHeight : this.activeBones?.height ?? 0;
        const capturedHeight = this.activeBones?.height ?? 0;
        return effectiveHeight > 0 && capturedHeight > 0 ? effectiveHeight / capturedHeight : 1;
    }
    ngAfterViewInit() {
        this.updateDarkMode();
        this.updateBones();
        const el = this.containerRef?.nativeElement;
        if (el) {
            const rect = el.getBoundingClientRect();
            this.containerWidth = Math.round(rect.width);
            if (rect.height > 0)
                this.containerHeight = Math.round(rect.height);
            this.resizeObserver = new ResizeObserver(entries => {
                const r = entries[0]?.contentRect;
                this.containerWidth = Math.round(r?.width ?? 0);
                if (r && r.height > 0)
                    this.containerHeight = Math.round(r.height);
                this.updateBones();
                this.cdr.markForCheck();
            });
            this.resizeObserver.observe(el);
        }
        this.mutationObserver = new MutationObserver(() => {
            this.updateDarkMode();
            this.cdr.markForCheck();
        });
        this.mutationObserver.observe(document.documentElement, {
            attributes: true,
            attributeFilter: ['class'],
        });
        this.mq = window.matchMedia('(prefers-color-scheme: dark)');
        this.mqHandler = () => { this.updateDarkMode(); this.cdr.markForCheck(); };
        this.mq.addEventListener('change', this.mqHandler);
        this.cdr.markForCheck();
    }
    ngOnChanges(changes) {
        if (changes['loading'] || changes['name'] || changes['initialBones']) {
            this.updateBones();
        }
        if (changes['loading']) {
            const prev = changes['loading'].previousValue;
            const curr = changes['loading'].currentValue;
            if (prev && !curr && this.transitionMs > 0 && this.activeBones) {
                if (this.transitionTimer)
                    clearTimeout(this.transitionTimer);
                this.transitioning = true;
                this.transitionTimer = setTimeout(() => {
                    this.transitioning = false;
                    this.transitionTimer = null;
                    this.cdr.markForCheck();
                }, this.transitionMs);
            }
        }
    }
    ngOnDestroy() {
        if (this.mq && this.mqHandler)
            this.mq.removeEventListener('change', this.mqHandler);
        this.mutationObserver?.disconnect();
        this.resizeObserver?.disconnect();
    }
    updateDarkMode() {
        if (typeof window === 'undefined')
            return;
        const el = this.containerRef?.nativeElement;
        this.isDark =
            document.documentElement.classList.contains('dark') ||
                !!el?.closest('.dark');
    }
    updateBones() {
        const effectiveBones = this.initialBones ?? (this.name ? getRegisteredBones(this.name) : undefined);
        const viewportWidth = typeof window !== 'undefined' ? window.innerWidth : this.containerWidth;
        const resolveWidth = this.containerWidth > 0 ? this.containerWidth : viewportWidth;
        this.activeBones = effectiveBones && resolveWidth > 0
            ? resolveResponsive(effectiveBones, resolveWidth)
            : null;
    }
    trackBone(index) {
        return index;
    }
    get visibleBones() {
        return (this.activeBones?.bones ?? []).filter(raw => !normalizeBone(raw).c);
    }
    getBoneStyle(raw, index = 0) {
        const bone = normalizeBone(raw);
        const radius = typeof bone.r === 'string' ? bone.r : `${bone.r}px`;
        const boneColor = this.resolvedColor;
        const capturedPxW = (bone.w / 100) * (this.activeBones?.width ?? 0);
        const isCircle = bone.r === '50%' && (this.activeBones?.width ?? 0) > 0 && Math.abs(capturedPxW - bone.h) < 4;
        const w = isCircle ? `${bone.h * this.scaleY}px` : `${bone.w}%`;
        const stagger = this.staggerMs > 0
            ? `opacity:0;animation:by-${this.uid} 0.3s ease-out ${index * this.staggerMs}ms forwards;`
            : '';
        return `position:absolute;left:${bone.x}%;top:${bone.y * this.scaleY}px;width:${w};height:${bone.h * this.scaleY}px;border-radius:${radius};background-color:${boneColor};overflow:hidden;${stagger}`;
    }
    getOverlayStyle() {
        const anim = this.animationStyle;
        if (anim === 'solid')
            return '';
        const lighterColor = adjustColor(this.resolvedColor, this.isDark ? PULSE.darkAdjust : PULSE.lightAdjust);
        if (anim === 'pulse') {
            return `position:absolute;inset:0;background-color:${lighterColor};animation:bp-${this.uid} ${PULSE.speed} ease-in-out infinite;`;
        }
        if (anim === 'shimmer') {
            return `position:absolute;inset:0;background:linear-gradient(${SHIMMER.angle}deg, transparent ${SHIMMER.start}%, ${lighterColor} 50%, transparent ${SHIMMER.end}%);background-size:200% 100%;animation:bs-${this.uid} ${SHIMMER.speed} linear infinite;`;
        }
        return '';
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "21.2.8", ngImport: i0, type: SkeletonComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "21.2.8", type: SkeletonComponent, isStandalone: true, selector: "boneyard-skeleton", inputs: { loading: "loading", name: "name", initialBones: "initialBones", color: "color", darkColor: "darkColor", animate: "animate", stagger: "stagger", transition: "transition", boneClass: "boneClass", cssClass: "cssClass", snapshotConfig: "snapshotConfig" }, viewQueries: [{ propertyName: "containerRef", first: true, predicate: ["container"], descendants: true }], usesOnChanges: true, ngImport: i0, template: `
    <!-- Build mode: render content for CLI capture -->
    <div
      *ngIf="buildMode; else runtimeMode"
      #container
      [class]="cssClass"
      style="position:relative;"
      [attr.data-boneyard]="name"
      [attr.data-boneyard-config]="serializedSnapshotConfig"
    >
      <div>
        <ng-content select="[fixture]"></ng-content>
        <ng-content></ng-content>
      </div>
    </div>

    <!-- Runtime mode -->
    <ng-template #runtimeMode>
      <div
        #container
        [class]="cssClass"
        style="position:relative;"
        [attr.data-boneyard]="name"
        [attr.data-boneyard-config]="serializedSnapshotConfig"
      >
        <div data-boneyard-content="true" [style.visibility]="showSkeleton && !transitioning ? 'hidden' : null">
          <ng-container *ngIf="showFallback; else defaultContent">
            <ng-content select="[fallback]"></ng-content>
          </ng-container>
          <ng-template #defaultContent>
            <ng-content></ng-content>
          </ng-template>
        </div>

        <div
          *ngIf="showSkeleton && activeBones"
          data-boneyard-overlay="true"
          [style]="'position:absolute;inset:0;overflow:hidden;opacity:' + (transitioning ? 0 : 1) + ';' + (transitionMs > 0 ? 'transition:opacity ' + transitionMs + 'ms ease-out;' : '')"
        >
          <div style="position:relative;width:100%;height:100%;">
            <div
              *ngFor="let bone of visibleBones; let i = index; trackBy: trackBone"
              data-boneyard-bone="true"
              [class]="resolvedBoneClass"
              [style]="getBoneStyle(bone, i)"
            >
              <div
                *ngIf="animationStyle !== 'solid'"
                [style]="getOverlayStyle()"
              ></div>
            </div>

            <style *ngIf="animationStyle === 'pulse'">
              @keyframes bp-{{ uid }} { 0%,100%{opacity:0} 50%{opacity:1} }
            </style>
            <style *ngIf="animationStyle === 'shimmer'">
              @keyframes bs-{{ uid }} { 0%{background-position:200% 0} 100%{background-position:-200% 0} }
            </style>
            <style *ngIf="staggerMs > 0">
              @keyframes by-{{ uid }} { from{opacity:0} to{opacity:1} }
            </style>
          </div>
        </div>
      </div>
    </ng-template>
  `, isInline: true, styles: ["\n              @keyframes bp-{{ uid }} { 0%,100%{opacity:0} 50%{opacity:1} }\n            ", "\n              @keyframes bs-{{ uid }} { 0%{background-position:200% 0} 100%{background-position:-200% 0} }\n            ", "\n              @keyframes by-{{ uid }} { from{opacity:0} to{opacity:1} }\n            "], dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "21.2.8", ngImport: i0, type: SkeletonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'boneyard-skeleton',
                    standalone: true,
                    imports: [CommonModule],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
    <!-- Build mode: render content for CLI capture -->
    <div
      *ngIf="buildMode; else runtimeMode"
      #container
      [class]="cssClass"
      style="position:relative;"
      [attr.data-boneyard]="name"
      [attr.data-boneyard-config]="serializedSnapshotConfig"
    >
      <div>
        <ng-content select="[fixture]"></ng-content>
        <ng-content></ng-content>
      </div>
    </div>

    <!-- Runtime mode -->
    <ng-template #runtimeMode>
      <div
        #container
        [class]="cssClass"
        style="position:relative;"
        [attr.data-boneyard]="name"
        [attr.data-boneyard-config]="serializedSnapshotConfig"
      >
        <div data-boneyard-content="true" [style.visibility]="showSkeleton && !transitioning ? 'hidden' : null">
          <ng-container *ngIf="showFallback; else defaultContent">
            <ng-content select="[fallback]"></ng-content>
          </ng-container>
          <ng-template #defaultContent>
            <ng-content></ng-content>
          </ng-template>
        </div>

        <div
          *ngIf="showSkeleton && activeBones"
          data-boneyard-overlay="true"
          [style]="'position:absolute;inset:0;overflow:hidden;opacity:' + (transitioning ? 0 : 1) + ';' + (transitionMs > 0 ? 'transition:opacity ' + transitionMs + 'ms ease-out;' : '')"
        >
          <div style="position:relative;width:100%;height:100%;">
            <div
              *ngFor="let bone of visibleBones; let i = index; trackBy: trackBone"
              data-boneyard-bone="true"
              [class]="resolvedBoneClass"
              [style]="getBoneStyle(bone, i)"
            >
              <div
                *ngIf="animationStyle !== 'solid'"
                [style]="getOverlayStyle()"
              ></div>
            </div>

            <style *ngIf="animationStyle === 'pulse'">
              @keyframes bp-{{ uid }} { 0%,100%{opacity:0} 50%{opacity:1} }
            </style>
            <style *ngIf="animationStyle === 'shimmer'">
              @keyframes bs-{{ uid }} { 0%{background-position:200% 0} 100%{background-position:-200% 0} }
            </style>
            <style *ngIf="staggerMs > 0">
              @keyframes by-{{ uid }} { from{opacity:0} to{opacity:1} }
            </style>
          </div>
        </div>
      </div>
    </ng-template>
  `,
                }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }], propDecorators: { loading: [{
                type: Input
            }], name: [{
                type: Input
            }], initialBones: [{
                type: Input
            }], color: [{
                type: Input
            }], darkColor: [{
                type: Input
            }], animate: [{
                type: Input
            }], stagger: [{
                type: Input
            }], transition: [{
                type: Input
            }], boneClass: [{
                type: Input
            }], cssClass: [{
                type: Input
            }], snapshotConfig: [{
                type: Input
            }], containerRef: [{
                type: ViewChild,
                args: ['container']
            }] } });
