// boneyard-js/dist/react.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useRef, useState, useEffect, useLayoutEffect } from 'react';
import { normalizeBone } from './types.js';
import { adjustColor, ensureBuildSnapshotHook, getRegisteredBones, isBuildMode, registerBones, resolveResponsive, SHIMMER, PULSE, DEFAULTS, } from './shared.js';
ensureBuildSnapshotHook();
export { registerBones };
let globalConfig = {};
/**
 * Set global defaults for all `<Skeleton>` components.
 * Individual props override these defaults.
 *
 * ```ts
 * import { configureBoneyard } from 'boneyard-js/react'
 *
 * configureBoneyard({
 *   color: '#e5e5e5',
 *   darkColor: '#2a2a2a',
 *   animate: true,
 * })
 * ```
 */
export function configureBoneyard(config) {
    globalConfig = { ...globalConfig, ...config };
}
/**
 * Wrap any component to get automatic skeleton loading screens.
 *
 * 1. Run `npx boneyard-js build` — captures bone positions from your rendered UI
 * 2. Import the generated registry in your app entry
 * 3. `<Skeleton name="..." loading={isLoading}>` auto-resolves bones by name
 */
export function Skeleton({ loading, children, name, initialBones, color, darkColor, animate, stagger = false, transition = false, boneClass, className, fallback, fixture, snapshotConfig, }) {
    const containerRef = useRef(null);
    const uid = useRef(Math.random().toString(36).slice(2, 8)).current;
    const [containerWidth, setContainerWidth] = useState(0);
    const [containerHeight, setContainerHeight] = useState(0);
    const [isDark, setIsDark] = useState(false);
    // Auto-detect dark mode via .dark class on <html> or ancestor
    useEffect(() => {
        if (typeof window === 'undefined')
            return;
        const checkDark = () => {
            const hasDarkClass = document.documentElement.classList.contains('dark') ||
                !!containerRef.current?.closest('.dark');
            setIsDark(hasDarkClass);
        };
        checkDark();
        // MutationObserver catches .dark toggling on <html>
        const mo = new MutationObserver(checkDark);
        mo.observe(document.documentElement, { attributes: true, attributeFilter: ['class'] });
        // matchMedia catches OS theme changes that may toggle .dark on non-<html> ancestors
        const mq = window.matchMedia('(prefers-color-scheme: dark)');
        mq.addEventListener('change', checkDark);
        return () => {
            mo.disconnect();
            mq.removeEventListener('change', checkDark);
        };
    }, []);
    const effectiveColor = color ?? globalConfig.color ?? DEFAULTS.web.light;
    const effectiveDarkColor = darkColor ?? globalConfig.darkColor ?? DEFAULTS.web.dark;
    const resolvedColor = isDark ? effectiveDarkColor : effectiveColor;
    const rawAnimate = animate ?? globalConfig.animate ?? 'pulse';
    const animationStyle = rawAnimate === true ? 'pulse' :
        rawAnimate === false ? 'solid' :
            rawAnimate;
    // Track container width for responsive breakpoint selection
    useEffect(() => {
        const el = containerRef.current;
        if (!el)
            return;
        const ro = new ResizeObserver(entries => {
            const rect = entries[0]?.contentRect;
            setContainerWidth(Math.round(rect?.width ?? 0));
            if (rect && rect.height > 0)
                setContainerHeight(Math.round(rect.height));
        });
        ro.observe(el);
        const rect = el.getBoundingClientRect();
        setContainerWidth(Math.round(rect.width));
        if (rect.height > 0)
            setContainerHeight(Math.round(rect.height));
        return () => ro.disconnect();
    }, []);
    // Data attributes for CLI discovery
    const dataAttrs = {};
    if (name) {
        dataAttrs['data-boneyard'] = name;
        if (snapshotConfig) {
            dataAttrs['data-boneyard-config'] = JSON.stringify(snapshotConfig);
        }
    }
    // Build mode: render fixture (if provided) or children so CLI can capture bones
    if (isBuildMode()) {
        return (_jsx("div", { ref: containerRef, className: className, style: { position: 'relative' }, ...dataAttrs, children: _jsx("div", { children: fixture ?? children }) }));
    }
    // Resolve bones: explicit initialBones > registry lookup
    // Use viewport width to pick breakpoint since bones are keyed by viewport width
    // After mount, use window.innerWidth as fallback so bones render immediately
    // without waiting for ResizeObserver. Before mount (SSR/hydration), use 0
    // to avoid hydration mismatch.
    const [mounted, setMounted] = useState(false);
    useLayoutEffect(() => { setMounted(true); }, []);
    const effectiveBones = initialBones ?? (name ? getRegisteredBones(name) : undefined);
    const viewportWidth = mounted && typeof window !== 'undefined' ? window.innerWidth : 0;
    const resolveWidth = containerWidth > 0 ? containerWidth : viewportWidth;
    const activeBones = effectiveBones && resolveWidth > 0
        ? resolveResponsive(effectiveBones, resolveWidth)
        : null;
    const resolvedBoneClass = boneClass ?? globalConfig.boneClass;
    // Stagger: delay between each bone's animation
    const staggerMs = (() => { const v = stagger ?? globalConfig.stagger; return v === true ? 80 : v === false || !v ? 0 : v; })();
    // Transition: fade out skeleton when loading ends
    const transitionMs = (() => { const v = transition ?? globalConfig.transition; return v === true ? 300 : v === false || !v ? 0 : v; })();
    const [transitioning, setTransitioning] = useState(false);
    const prevLoadingRef = useRef(loading);
    const transitionTimerRef = useRef(null);
    useEffect(() => {
        if (prevLoadingRef.current && !loading && transitionMs > 0 && activeBones) {
            if (transitionTimerRef.current)
                clearTimeout(transitionTimerRef.current);
            setTransitioning(true);
            transitionTimerRef.current = setTimeout(() => {
                setTransitioning(false);
                transitionTimerRef.current = null;
            }, transitionMs);
        }
        prevLoadingRef.current = loading;
        return () => {
            if (transitionTimerRef.current)
                clearTimeout(transitionTimerRef.current);
        };
    }, [loading, transitionMs, activeBones]);
    const showSkeleton = (loading || transitioning) && activeBones;
    const showFallback = loading && !activeBones && !transitioning;
    // Scale vertical positions to match actual container height
    const effectiveHeight = containerHeight > 0 ? containerHeight : activeBones?.height ?? 0;
    const capturedHeight = activeBones?.height ?? 0;
    const scaleY = (effectiveHeight > 0 && capturedHeight > 0) ? effectiveHeight / capturedHeight : 1;
    return (_jsxs("div", { ref: containerRef, className: className, style: { position: 'relative' }, ...dataAttrs, children: [_jsx("div", { "data-boneyard-content": "true", style: showSkeleton && !transitioning ? { visibility: 'hidden' } : undefined, children: showFallback ? fallback : children }), showSkeleton && (_jsx("div", { "data-boneyard-overlay": "true", style: {
                    position: 'absolute', inset: 0, overflow: 'hidden',
                    opacity: transitioning ? 0 : 1,
                    transition: transitionMs > 0 ? `opacity ${transitionMs}ms ease-out` : undefined,
                }, children: _jsxs("div", { style: { position: 'relative', width: '100%', height: '100%' }, children: [activeBones.bones.filter(raw => !normalizeBone(raw).c).map((raw, i) => {
                            const b = normalizeBone(raw);
                            const boneColor = resolvedColor;
                            const lighterColor = adjustColor(resolvedColor, isDark ? PULSE.darkAdjust : PULSE.lightAdjust);
                            const capturedPxW = (b.w / 100) * (activeBones.width ?? 0);
                            const isCircle = b.r === '50%' && Math.abs(capturedPxW - b.h) < 4;
                            const boneStyle = {
                                position: 'absolute',
                                left: `${b.x}%`,
                                top: b.y * scaleY,
                                width: isCircle ? b.h * scaleY : `${b.w}%`,
                                height: b.h * scaleY,
                                borderRadius: typeof b.r === 'string' ? b.r : `${b.r}px`,
                                backgroundColor: boneColor,
                            };
                            const effectiveSpeed = globalConfig.speed;
                            if (animationStyle === 'pulse') {
                                boneStyle.animation = `bp-${uid} ${effectiveSpeed ?? PULSE.speed} ease-in-out infinite`;
                            }
                            else if (animationStyle === 'shimmer') {
                                const shimmerHighlight = isDark
                                    ? (globalConfig.darkShimmerColor ?? SHIMMER.darkHighlight)
                                    : (globalConfig.shimmerColor ?? SHIMMER.lightHighlight);
                                const angle = globalConfig.shimmerAngle ?? SHIMMER.angle;
                                delete boneStyle.backgroundColor;
                                boneStyle.backgroundImage = `linear-gradient(${angle}deg, ${boneColor} ${SHIMMER.start}%, ${shimmerHighlight} 50%, ${boneColor} ${SHIMMER.end}%)`;
                                boneStyle.backgroundSize = '200% 100%';
                                boneStyle.animation = `bs-${uid} ${effectiveSpeed ?? SHIMMER.speed} linear infinite`;
                            }
                            if (staggerMs > 0) {
                                boneStyle.opacity = 0;
                                boneStyle.animation = `${boneStyle.animation ? boneStyle.animation + ',' : ''} by-${uid} 0.3s ease-out ${i * staggerMs}ms forwards`;
                            }
                            return _jsx("div", { "data-boneyard-bone": "true", className: resolvedBoneClass, style: boneStyle }, i);
                        }), animationStyle === 'pulse' && (_jsx("style", { children: `@keyframes bp-${uid}{0%,100%{background-color:${resolvedColor}}50%{background-color:${adjustColor(resolvedColor, isDark ? PULSE.darkAdjust : PULSE.lightAdjust)}}}` })), animationStyle === 'shimmer' && (_jsx("style", { children: `@keyframes bs-${uid}{0%{background-position:200% 0}100%{background-position:-200% 0}}` })), staggerMs > 0 && (_jsx("style", { children: `@keyframes by-${uid}{from{opacity:0}to{opacity:1}}` }))] }) }))] }));
}
